src/Pipeline/Atomic.php

```diff
@@ -62,56 +64,82 @@ protected function getConnection()
 
-            if ($response instanceof ErrorResponseInterface) {
-                $connection->executeCommand($profile->createCommand('discard'));
-                throw new ServerException($response->getMessage());
+            // throw first error as a ServerException
+            if (($response instanceof ErrorResponseInterface) && !$ex) {
+                $ex = new ServerException($response->getMessage());
             }
         }
 
-        $executed = $connection->executeCommand($profile->createCommand('exec'));
+        // response to EXEC
+        $responses = $connection->readResponse($exec);
 
-        if (!isset($executed)) {
-            // TODO: should be throwing a more appropriate exception.
-            throw new ClientException(
-                'The underlying transaction has been aborted by the server.'
-            );
+        // if one of the commands returned an error, only throw after all responses have been received
+        if ($ex) {
+            throw $ex;
         }
 
-        if (count($executed) !== count($commands)) {
+        // if EXEC returned null, the transaction was aborted
+        if (!isset($responses)) {
+            // fake a MultiExec, which is not used in an atomic pipeline
+            $multi_exec = new MultiExec($this->getClient());
+            throw new AbortedMultiExecException($multi_exec, 'The underlying transaction has been aborted by the server.');
+        }
+
+        // if EXEC returned an error, throw ServerException (but don't close connection, all responses
+        // have been read and are accounted for)
+        if ($responses instanceof ErrorResponseInterface) {
+            throw new ServerException($responses->getMessage());
+        }
+
+        if (count($responses) !== count($commands)) {
             $expected = count($commands);
-            $received = count($executed);
+            $received = count($responses);
 
             throw new ClientException(
                 "Invalid number of responses [expected $expected, received $received]."
             );
         }
 
-        $responses = array();
-        $sizeOfPipe = count($commands);
         $exceptions = $this->throwServerExceptions();
 
-        for ($i = 0; $i < $sizeOfPipe; ++$i) {
-            $command = $commands->dequeue();
-            $response = $executed[$i];
+        // parse all unparsed results
+        $i = 0;
+        foreach ($commands as $command) {
+            $response = $responses[$i];
 
             if (!$response instanceof ResponseInterface) {
-                $responses[] = $command->parseResponse($response);
+                $responses[$i] = $command->parseResponse($response);
             } elseif ($response instanceof ErrorResponseInterface && $exceptions) {
                 $this->exception($connection, $response);
-            } else {
-                $responses[] = $response;
             }
 
-            unset($executed[$i]);
```

Comment:
Same as above: the reason for this is to free as much memory as possible early on during the iteration.

Reply:
(as with above)

Reply:
(reverted this as well.)